File: graphs/find_if_path_exists_in_graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        if (edges.size() == 0) return true;

        unordered_map<int, vector<int>> graph;
        for (int i = 0; i < edges.size(); ++i) {
            graph[edges[i][0]].push_back(edges[i][1]);
            graph[edges[i][1]].push_back(edges[i][0]);
        }

        unordered_set<int> visited;
        queue<int> q;
        q.push(source);

        while (!q.empty()) {
            int current = q.front();
            q.pop();
            if (current == destination) return true;

            for (int neighbor : graph[current]) {
                if (!visited.contains(neighbor)) {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }

        return false;
    }
};

File: graphs/find_if_path_exists_in_graph.cs
public class Solution {
    public bool ValidPath(int n, int[][] edges, int source, int destination) {
        if (edges.Length == 0 && n == 1) return true;

        Dictionary<int, List<int>> graph = [];
        for (int i = 0; i < edges.Length; ++i) {
            int u = edges[i][0];
            int v = edges[i][1];

            if (!graph.ContainsKey(u)) graph[u] = [];
            if (!graph.ContainsKey(v)) graph[v] = [];

            graph[u].Add(v);
            graph[v].Add(u);
        }

        HashSet<int> visited = [];
        Queue<int> q = [];
        q.Enqueue(source);
        while (q.TryDequeue(out int current)) {
            if (current == destination) return true;
            foreach (int neighbor in graph[current]) {
                if (visited.Add(neighbor)) q.Enqueue(neighbor);
            }
        }

        return false;
    }
}
